File: android/app/src/main/kotlin/com/orialthq/ori_beauty/share/IncomingImageSignature.kt
package com.orialthq.ori_beauty.share

internal data class IncomingImageFormat(
    val mimeType: String,
    val extension: String,
)

internal object IncomingImageSignature {
    private val jpeg = IncomingImageFormat("image/jpeg", "jpg")
    private val png = IncomingImageFormat("image/png", "png")
    private val webp = IncomingImageFormat("image/webp", "webp")

    fun detect(header: ByteArray): IncomingImageFormat? {
        if (
            header.size >= 3 &&
            header[0].unsigned() == 0xff &&
            header[1].unsigned() == 0xd8 &&
            header[2].unsigned() == 0xff
        ) {
            return jpeg
        }

        if (
            header.size >= 8 &&
            header[0].unsigned() == 0x89 &&
            header[1].unsigned() == 0x50 &&
            header[2].unsigned() == 0x4e &&
            header[3].unsigned() == 0x47 &&
            header[4].unsigned() == 0x0d &&
            header[5].unsigned() == 0x0a &&
            header[6].unsigned() == 0x1a &&
            header[7].unsigned() == 0x0a
        ) {
            return png
        }

        if (
            header.size >= 12 &&
            header.copyOfRange(0, 4).decodeToString() == "RIFF" &&
            header.copyOfRange(8, 12).decodeToString() == "WEBP"
        ) {
            return webp
        }

        return null
    }

    private fun Byte.unsigned(): Int = toInt() and 0xff
}

File: ios/Runner/Share/IncomingImageSignature.swift
import Foundation

struct IncomingImageFormat {
  let mimeType: String
  let fileExtension: String
}

/// Byte-for-byte mirror of `share/IncomingImageSignature.kt`. Both platforms must
/// accept and reject exactly the same headers, so keep the two in sync.
enum IncomingImageSignature {
  private static let jpeg = IncomingImageFormat(mimeType: "image/jpeg", fileExtension: "jpg")
  private static let png = IncomingImageFormat(mimeType: "image/png", fileExtension: "png")
  private static let webp = IncomingImageFormat(mimeType: "image/webp", fileExtension: "webp")

  static func detect(header: [UInt8]) -> IncomingImageFormat? {
    if header.count >= 3,
      header[0] == 0xff,
      header[1] == 0xd8,
      header[2] == 0xff
    {
      return jpeg
    }

    if header.count >= 8,
      header[0] == 0x89,
      header[1] == 0x50,
      header[2] == 0x4e,
      header[3] == 0x47,
      header[4] == 0x0d,
      header[5] == 0x0a,
      header[6] == 0x1a,
      header[7] == 0x0a
    {
      return png
    }

    if header.count >= 12,
      String(decoding: header[0..<4], as: UTF8.self) == "RIFF",
      String(decoding: header[8..<12], as: UTF8.self) == "WEBP"
    {
      return webp
    }

    return nil
  }
}
